app/features/receive/receive-cashu-token-hooks.tsx

```diff
@@ -0,0 +1,354 @@
+import type {
+  Account,
+  CashuAccount,
+  ExtendedCashuAccount,
+} from '~/features/accounts/account';
+import {
+  useAccount,
+  useAccounts,
+  useDefaultAccount,
+} from '~/features/accounts/account-hooks';
+import { tokenToMoney } from '~/features/shared/cashu';
+import {
+  getP2PKPubkeyFromProofs,
+  getUnspentProofsFromToken,
+  isP2PKSecret,
+  isPlainSecret,
+} from '~/lib/cashu';
+import { checkIsTestMint, getMintInfo } from '~/lib/cashu';
+import { useLatest } from '~/lib/use-latest';
+import type { AccountWithBadges } from '../accounts/account-selector';
+import {
+  useCreateCashuTokenSwap,
+  useTokenSwap,
+} from './cashu-token-swap-hooks';
+
+type CashuAccountWithBadges = AccountWithBadges<CashuAccount>;
+
+type UseGetClaimableTokenProps = {
+  token: Token;
+  cashuPubKey: string;
+};
+
+type TokenQueryResult =
+  | {
+      /** The token with only claimable proofs. Will be null if the token cannot be claimed */
+      claimableToken: null;
+      /** The reason why the token cannot be claimed. Will always be defined when claimableToken is null */
+      cannotClaimReason: string;
+    }
+  | {
+      claimableToken: Token;
+      cannotClaimReason: null;
+    };
+
+/**
+ * Takes a token and returns the account that the token is from.
+ * If the account does not exist, we construct and return an account, but we do not store it in the database.
+ */
+export function useGetCashuTokenSourceAccount(token: Token) {
+  const tokenCurrency = tokenToMoney(token).currency;
+  const { data: allAccounts } = useAccounts({ type: 'cashu' });
+  const existingAccount = allAccounts.find(
+    (a) => a.mintUrl === token.mint && a.currency === tokenCurrency,
+  );
+
+  if (existingAccount) {
+    return existingAccount;
+  }
+
+  const { data } = useSuspenseQuery({
+    queryKey: ['token-source-account', token.mint, tokenCurrency],
+    queryFn: async (): Promise<ExtendedCashuAccount> => {
+      const info = await getMintInfo(token.mint);
+      const isTestMint = await checkIsTestMint(token.mint);
+      return {
+        id: '',
+        type: 'cashu',
+        mintUrl: token.mint,
+        createdAt: new Date().toISOString(),
+        name: info?.name ?? 'Unknown Mint',
+        currency: tokenToMoney(token).currency,
+        isTestMint: isTestMint ?? false,
+        version: 0,
+        keysetCounters: {},
+        proofs: [],
+        isDefault: false,
+      };
+    },
+    staleTime: 3 * 60 * 1000,
+    retry: 1,
+  });
+
+  return data;
+}
+
+const getClaimableProofs = (unspentProofs: Proof[], cashuPubKey: string) => {
+  let cannotClaimReason: string | undefined;
+  const claimableProofs = unspentProofs.filter((proof) => {
+    if (isPlainSecret(proof.secret)) {
+      return true;
+    }
+    if (!isP2PKSecret(proof.secret)) {
+      cannotClaimReason = 'This ecash contains an unknown spending condition.';
+      return false;
+    }
+    try {
+      const pubkey = getP2PKPubkeyFromProofs([proof]);
```

Comment:
nit but I would maybe change getP2PKPubkeyFromProofs to return result in a way that we don't have to handle exceptions. I find that cleaner than having to write try catch